astoria/managers/astmetad.py

```diff
@@ -0,0 +1,150 @@
+    loop.run_until_complete(metad.run())
+
+
+class MetadataManager(DiskHandlerMixin, StateManager[MetadataManagerMessage]):
+    """Astoria Metadata State Manager."""
+
+    name = "astmetad"
+    dependencies = ["astdiskd"]
+
+    def _init(self) -> None:
+        self._lifecycle: Optional[MetadataDiskLifecycle] = None
+        self._requested_data: Dict[str, str] = {}
+
+        self._cur_disks: Dict[DiskUUID, DiskInfo] = {}
+        self._mqtt.subscribe("astdiskd", self.handle_astdiskd_disk_info_message)
+
+    @property
+    def offline_status(self) -> MetadataManagerMessage:
+        """
+        Status to publish when the manager goes offline.
+
+        This status should ensure that any other components relying
+        on this data go into a safe state.
+        """
+        return MetadataManagerMessage(
+            status=MetadataManagerMessage.Status.STOPPED,
+            metadata=Metadata.init(self.config),
+        )
+
+    async def main(self) -> None:
+        """Main routine for astmetad."""
+        self.update_status()
+
+        # Wait whilst the program is running.
+        await self.wait_loop()
+
+        for uuid, info in self._cur_disks.items():
+            asyncio.ensure_future(self.handle_disk_removal(uuid, info))
+
+    async def handle_disk_insertion(self, uuid: DiskUUID, disk_info: DiskInfo) -> None:
+        """Handle a disk insertion."""
+        LOGGER.debug(f"Disk inserted: {uuid} ({disk_info.disk_type})")
+        if disk_info.disk_type is DiskType.METADATA:
+            LOGGER.info(f"Metadata disk {uuid} is mounted at {disk_info.mount_path}")
+            if self._lifecycle is None:
+                LOGGER.debug(f"Starting metadata lifecycle for {uuid}")
+                self._lifecycle = MetadataDiskLifecycle(uuid, disk_info)
+                self.update_status()
+            else:
+                LOGGER.warn("Cannot use metadata, there is already a lifecycle present.")
+
+    async def handle_disk_removal(self, uuid: DiskUUID, disk_info: DiskInfo) -> None:
+        """Handle a disk removal."""
+        LOGGER.debug(f"Disk removed: {uuid} ({disk_info.disk_type})")
+        if disk_info.disk_type is DiskType.METADATA:
+            LOGGER.info(f"Metadata disk {uuid} removed ({disk_info.mount_path})")
+            if self._lifecycle is not None and self._lifecycle._uuid == disk_info.uuid:
+                self._lifecycle = None
+                self.update_status()
+
+    def get_current_metadata(self) -> Metadata:
+        """
+        Calculate the current metadata.
+
+        Takes the default, static metadata based on the config and system
+        information. It then overlays data from other sources in a priority order,
+        whereby each source has a set of permitted attributes in the metadata that
+        can be overriden.
+        """
+        # Mutation data sources in priority order.
+        MUTATION_SOURCES: List[Tuple[Set[str], Dict[str, str]]] = [
```

Comment:
why the `ALL_CAPS` if it's not a constant, ooi?

Reply:
A mistake leftover from when it was a constant earlier in development of this PR. Will fix.

Reply:
6037684